libs/utils/src/pg_checksum_page.rs

```diff
@@ -0,0 +1,120 @@
+///
+///  Port of Postgres pg_checksum_page
+///
+
+const BLCKSZ: usize = 8192;
+const N_SUMS: usize = 32;
+/* prime multiplier of FNV-1a hash */
+const FNV_PRIME: u32 = 16777619;
+
+/*
+ * Base offsets to initialize each of the parallel FNV hashes into a
+ * different initial state.
+ */
+const CHECKSUM_BASE_OFFSETS: [u32; N_SUMS] = [
+    0x5B1F36E9, 0xB8525960, 0x02AB50AA, 0x1DE66D2A, 0x79FF467A, 0x9BB9F8A3, 0x217E7CD2, 0x83E13D2C,
+    0xF8D4474F, 0xE39EB970, 0x42C6AE16, 0x993216FA, 0x7B093B5D, 0x98DAFF3C, 0xF718902A, 0x0B1C9CDB,
+    0xE58F764B, 0x187636BC, 0x5D7B3BB1, 0xE73DE7DE, 0x92BEC979, 0xCCA6C0B2, 0x304A0979, 0x85AA43D4,
+    0x783125BB, 0x6CA8EAA2, 0xE407EAC6, 0x4B5CFC3E, 0x9FBF8C76, 0x15CA20BE, 0xF2CA9FD3, 0x959BD756,
+];
+
+/*
+ * Calculate one round of the checksum.
+ */
+fn checksum_comp(checksum: u32, value: u32) -> u32 {
+    let tmp = checksum ^ value;
+    tmp.wrapping_mul(FNV_PRIME) ^ (tmp >> 17)
+}
+
+/*
+ * Compute the checksum for a Postgres page.
+ *
+ * The page must be adequately aligned (at least on a 4-byte boundary).
+ * Beware also that the checksum field of the page is transiently zeroed.
+ *
+ * The checksum includes the block number (to detect the case where a page is
+ * somehow moved to a different location), the page header (excluding the
+ * checksum itself), and the page data.
+ *
+ * As in C implementation in Postgres, the checksum attribute on the page is
+ * excluded from the calculation and preserved.
+ */
+pub fn pg_checksum_page(data: &[u8], blkno: u32) -> u16 {
+    let page = unsafe { std::mem::transmute::<&[u8], &[u32]>(data) };
+    let mut checksum: u32 = 0;
+    let mut sums = CHECKSUM_BASE_OFFSETS;
+
+    /* main checksum calculation */
+    for i in 0..(BLCKSZ / (4 * N_SUMS)) {
+        for (j, sum) in sums.iter_mut().enumerate().take(N_SUMS) {
+            let chunk_i = i * N_SUMS + j;
+            let chunk: u32;
+            if chunk_i == 2 {
+                let mut chunk_copy = page[chunk_i].to_le_bytes();
+                chunk_copy[0] = 0;
+                chunk_copy[1] = 0;
+                chunk = u32::from_le_bytes(chunk_copy);
```

Comment:
This PR has seen a handful of comments that revolve around various nits and styling issues, but this particular piece of code is very broken in multiple ways.

First of all, why not just

```
value && 0xFFFF0000
```

instead of this verbosity? The original code never said anything about endianness or LE in particular, but this striking detail drew my attention to the broader question: is this even correct?

Speaking of the original postgres code, it's very different. It uses `PageHeaderData` for resetting the checksum, which is hacky but still correct. But here we're taking it for granted that the lower half (I omit the to-le/from-le part) of a **certain** u32 will contain `pd_checksum`, **which is not the case for big-endian!**

Okay, I must admit that we'll probably never want to run our code on a big-endian arch nowadays, but we really shouldn't let such things slip in our code base unnoticed, without even a tiny comment clarifying that only little-endian is supported.

Also there's only a hint that this code has been adopted from postgres (like, you could grep for `pg_checksum_page`, but you'll soon learn that it's very different, like I did). In cases like this I'd love to see a comment saying _"this code has **not** been copied verbatim from `<original function name OR github URL>`"_.

So here's a proof that mixing field- and ptr-to-u32-based access patterns is a bad idea:

```c
#include <stdint.h>
#include <stdio.h>

struct Data {
   // Desugared PageXLogRecPtr
    uint32_t a;
    uint32_t b;

    // pd_checksum
    uint16_t c;
    
    // pd_flags
    uint16_t d;
};

int main() {
    struct Data data = (struct Data) {
        .a = 1,
        .b = 2,
        .c = 3,
        .d = 4,
    };

    uint32_t *ptr = (uint32_t *) &data;
    ptr[2] &= 0xffff0000;

    printf("%d\n", ((volatile struct Data *)&data)->c);
    return 0;
}
```

I picked a big-endian mips for the demonstation. For the cross-tollchain I'm using the [zig](https://github.com/ziglang/zig) compiler because it's super-convenient (I highly recommend it). We'll also need to install a pack of qemu-user interpreters (e.g. https://packages.debian.org/en/sid/qemu-user-static) to run the actual binaries.

Now, check this out:

```
$ zig cc test.c && ./a.out 
is little endian: 1
0

$ zig cc test.c --target=mips-linux-musl && ./a.out
is little endian: 0
3
```

Oops!

Reply:
> but this particular piece of code is very broken in multiple ways.

Thanks for the looking on it and such a detailed comment! Assuming that we are not running any of our service components on big-endian archs and, yes, likely not going to, what are the multiple ways we are talking about? Sorry if I missed something in your comment. Because yes, it certainly not intended to be run on big-endian and worth marking / commenting.

Reply:
It is not the only place in Neon code where we assume that target system is little-endian. May be it is not good, but I do not think that it is so critical at this moment.
It is possible now to find big-endian machine, but not in cloud and I do not see any reasons why somebody wants to run Neon on such system.